presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSourceSql.java

```diff
@@ -0,0 +1,149 @@
+    {
+        String sql = pinotQuery.getQuery();
+        JsonNode jsonBody;
+        try {
+            jsonBody = objectMapper.readTree(responseJsonString);
+        }
+        catch (IOException e) {
+            throw new PinotException(PINOT_UNEXPECTED_RESPONSE, Optional.of(sql), "Couldn't parse response", e);
+        }
+        handleCommonResponse(sql, jsonBody);
+        JsonNode resultTable = jsonBody.get("resultTable");
+        if (resultTable != null) {
+            JsonNode dataSchema = resultTable.get("dataSchema");
+            if (dataSchema == null) {
+                throw new PinotException(
+                    PINOT_UNEXPECTED_RESPONSE,
+                    Optional.of(sql),
+                    String.format("Expected data schema in the response"));
+            }
+            JsonNode columnDataTypes = dataSchema.get("columnDataTypes");
+            JsonNode columnNames = dataSchema.get("columnNames");
+
+            if (columnDataTypes == null
+                    || !columnDataTypes.isArray()
+                    || columnDataTypes.size() < blockBuilders.size()) {
+                throw new PinotException(
+                    PINOT_UNEXPECTED_RESPONSE,
+                    Optional.of(sql),
+                    String.format("ColumnDataTypes and results expected for %s, expected %d columnDataTypes but got %d", sql, blockBuilders.size(), columnDataTypes == null ? 0 : columnDataTypes.size()));
+            }
+            if (columnNames == null
+                    || !columnNames.isArray()
+                    || columnNames.size() < blockBuilders.size()) {
+                throw new PinotException(
+                    PINOT_UNEXPECTED_RESPONSE,
+                    Optional.of(sql),
+                    String.format("ColumnNames and results expected for %s, expected %d columnNames but got %d", sql, blockBuilders.size(), columnNames == null ? 0 : columnNames.size()));
+            }
+
+            JsonNode rows = resultTable.get("rows");
+            checkState(rows.size() >= 1, "Expected at least one row to be present");
+            setRows(sql, blockBuilders, types, rows);
+            return rows.size();
+        }
+        else {
+            throw new PinotException(
+                    PINOT_UNEXPECTED_RESPONSE,
+                    Optional.of(sql),
+                    "Expected resultTable to be present");
+        }
+    }
+
+    @VisibleForTesting
+    @Override
+    public BlockAndTypeBuilder buildBlockAndTypeBuilder(List<PinotColumnHandle> columnHandles,
+            GeneratedPinotQuery brokerSql)
+    {
+        List<Type> columnTypes = columnHandles.stream()
+                .map(PinotColumnHandle::getDataType)
+                .collect(Collectors.toList());
+        PageBuilder pageBuilder = new PageBuilder(columnTypes);
+        ArrayList<BlockBuilder> columnBlockBuilders = new ArrayList<>();
+        // Sql mode requires no need for expected column indices as Pinot Sql responses follow
```

Comment:
nit: I think we don't need to retain the historical reference to column indices. Instead should we simply say that the SQL broker page source returns the columns in the order requested by Presto.

Reply:
I think so. I will update this logic.

Reply:
done.